src/filters/WorldFilter.cpp

```diff
@@ -0,0 +1,129 @@
+//
+// Created by kjhertenberg on 13-5-19.
+//
+
+#include "roboteam_proto/messages_robocup_ssl_detection.pb.h"
+#include <filters/WorldFilter.h>
+
+namespace world {
+
+    WorldFilter::WorldFilter() {
+        blueBots.clear();
+        yellowBots.clear();
+        balls.clear();
+    }
+
+// if we get a new frame we update our observations
+    void WorldFilter::addFrame(const proto::SSL_DetectionFrame &msg) {
+        const double filterGrabDistance = 0.5;
+        double timeCapture = msg.t_capture();
+        uint cameraID = msg.camera_id();
+        handleRobots(yellowBots,msg.robots_yellow(), filterGrabDistance, timeCapture, cameraID);
+        handleRobots(blueBots,msg.robots_blue(), filterGrabDistance, timeCapture, cameraID);
+        for (const proto::SSL_DetectionBall &detBall : msg.balls()) {
+            bool addedBall = false;
+            for (const auto &filter : balls) {
+                if (filter->distanceTo(detBall.x(), detBall.y()) < filterGrabDistance) {
+                    filter->addObservation(detBall, timeCapture, cameraID);
+                    addedBall = true;
+                }
+            }
+            if (!addedBall) {
+                balls.push_back(std::make_unique<BallFilter>(detBall, timeCapture, cameraID));
+            }
+        }
+    }
+    void
+    WorldFilter::handleRobots(robotMap &robots,
+                              const google::protobuf::RepeatedPtrField<proto::SSL_DetectionRobot> &observations,
+                              double filterGrabDistance, double timeCapture, uint cameraID) {
+        for (const proto::SSL_DetectionRobot &robot : observations) {
+            bool addedBot = false;
+            for (const auto &filter : robots[robot.robot_id()]) {
+                if (filter->distanceTo(robot.x(), robot.y()) < filterGrabDistance) {
+                    filter->addObservation(robot, timeCapture, cameraID);
+                    addedBot = true;
+                }
+            }
+            if (!addedBot) {
+                // We create a new filter if no filter close to the robot exists
+                robots[robot.robot_id()].push_back(std::make_unique<RobotFilter>(robot, timeCapture, cameraID));
+            }
+        }
+    }
+
+//Creates a world message with the currently observed objects in it
+    proto::World WorldFilter::getWorld(double time) {
+        //First we update to the time we want packets at. Very important!
+        update(time, true);
+        proto::World world;
+        world.set_time(time);
+        for (const auto &kalmanYellowBotsOneId : yellowBots) {
```

Comment:
iterator "kalmanYellowBotsOneId" is named "filtersAndId" in updateRobots(). Should both have the same name